src/System.Memory/src/System/SpanExtensions.cs

```diff
@@ -274,5 +274,107 @@ public static void CopyTo<T>(this T[] array, Span<T> destination)
+        /// Determines whether two sequences overlap.
+        /// </summary>
+        public static unsafe bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second)
+        {
+            if (second.IsEmpty)
+                return false;
+
+            ref T firstRef = ref first.DangerousGetPinnableReference();
+            ref T secondRef = ref second.DangerousGetPinnableReference();
+
+            IntPtr firstByteCount = Unsafe.ByteOffset(ref firstRef, ref Unsafe.Add(ref firstRef, first.Length));
+            IntPtr secondByteCount = Unsafe.ByteOffset(ref secondRef, ref Unsafe.Add(ref secondRef, second.Length));
+
+            IntPtr diff = Unsafe.ByteOffset(ref firstRef, ref secondRef);
+
+            return (sizeof(IntPtr) == sizeof(int))
+                ? (uint)diff < (uint)firstByteCount || (uint)diff > (uint)-(int)secondByteCount
+                : (ulong)diff < (ulong)firstByteCount || (ulong)diff > (ulong)-(long)secondByteCount;
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap and outputs the element offset.
+        /// </summary>
+        public static unsafe bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            if (second.IsEmpty)
+            {
+                elementOffset = 0;
+                return false;
+            }
+
+            ref T firstRef = ref first.DangerousGetPinnableReference();
+            ref T secondRef = ref second.DangerousGetPinnableReference();
+
+            IntPtr firstByteCount = Unsafe.ByteOffset(ref firstRef, ref Unsafe.Add(ref firstRef, first.Length));
+            IntPtr secondByteCount = Unsafe.ByteOffset(ref secondRef, ref Unsafe.Add(ref secondRef, second.Length));
+
+            IntPtr diff = Unsafe.ByteOffset(ref firstRef, ref secondRef);
+
+            bool overlaps = (sizeof(IntPtr) == sizeof(int))
+                ? (uint)diff < (uint)firstByteCount || (uint)diff > (uint)-(int)secondByteCount
+                : (ulong)diff < (ulong)firstByteCount || (ulong)diff > (ulong)-(long)secondByteCount;
+
+            elementOffset = (sizeof(IntPtr) == sizeof(int))
+                ? overlaps ? (int)diff / Unsafe.SizeOf<T>() : 0
+                : overlaps ? (int)((long)diff / Unsafe.SizeOf<T>()) : 0;
+
+            return overlaps;
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap.
+        /// </summary>
+        public static unsafe bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second)
+        {
+            if (second.IsEmpty)
+                return false;
+
+            ref T firstRef = ref first.DangerousGetPinnableReference();
+            ref T secondRef = ref second.DangerousGetPinnableReference();
+
+            IntPtr firstByteCount = Unsafe.ByteOffset(ref firstRef, ref Unsafe.Add(ref firstRef, first.Length));
+            IntPtr secondByteCount = Unsafe.ByteOffset(ref secondRef, ref Unsafe.Add(ref secondRef, second.Length));
```

Comment:
didn't we conclude last time that only *one* `ByteOffset` call was necessary? and also not two `Add`s? however, optimization can wait until tests are in place.

Reply:
There is only one `ByteOffset` call that compares the two `ref`s. The other two `ByteOffset` calls are for calculating `first` and `second.Length * Unsafe.Sizeof<T>()` without having an `?:` for 32/64-bit.